chore: migrate release process to Ship.js

aseure opened the pull request:
chore: migrate release process to Ship.js

This commit introduces a new release process by:

- removing the need for the scripts/release.sh bash script in favor of
  https://github.com/algolia/shipjs.
- automating the release in CircleCI

Notable changes are:

- `version` attribute added to the `package.json` file: this is a bit
  redundant with the `version` attribute available in the `lerna.json`
  but it is required by `conventional-changelog-cli`, which is used by Ship.js,
  to generate new Changelog entries, and can only identify versions based
  on the `package.json` file.
- `CHANGELOG.md` text header has been removed because
  `conventional-changelog-ci` is not aware of headers in Changelog
  files.
- `GITHUB_TOKEN` environment variable is now added to the CircleCI
  configuration. It grants CircleCI with the appropriate rights to run
  `shipjs trigger`, effectively publishing to NPM. The value is set to
  the Github account `algolia-api-client-bot` which can be retrieved
  from our Vault cluster, like other credentials.
- `CONTRIBUTING.md` file updated to reflect the updated release process.

eunjae-lee commented on ship.config.js:
```diff
@@ -0,0 +1,30 @@
+  getTagName: ({ version }) => `${version}`,
+  conventionalChangelogArgs:
+    '--config conventional-changelog.config.js --infile CHANGELOG.md --same-file',
+  versionUpdated({ version, dir }) {
```
You're gonna need to update the version of the the root `package.json` here.

aseure replied:
This is already done by the `monerepo.mainVersionFile` set to `package.json` on line 9, along with the update of others package-specific `package.json` files.

Here in `versionUpdated`, I'm only manually updating the root `lerna.json` and `packages/client-common/src/version.ts` TypeScript file which aren't done automatically.

eunjae-lee commented:
You may want to address in `CONTIRBUTING.md` about how you should prepend `GITHUB_TOKEN` or add it to `.env` file before running `npm run release`.

aseure commented:
@eunjae-lee thanks for the review buddy and for your help with Ship.js! I've addressed your comments.

> You may want to address in CONTIRBUTING.md about how you should prepend GITHUB_TOKEN or add it to .env file before running npm run release.

Good point, just added it to the CONTRIBUTING.md file 👍

eunjae-lee reviewed (dismissed):
The change looks good to me, but I see a circleci job failing and some other jobs are not triggered yet.

aseure commented:
@Haroenv @eunjae-lee The `Required` checks are pending because I've had to explicitely rename the CircleCI jobs for this to pass. Are you fine if I edit the branch protection rules on my own right now or do you prefer to do it on your own later (because they are some pending pull requests opened as well).

I've updated the Confluence documentation and updated the CONTRIBUTING.md to reflect this, added our credentials to Vault. Could you have a last review before I merge this? Thanks!

Haroenv commented on .circleci/config.yml:
```diff
@@ -119,27 +119,94 @@ jobs:
+            - test_browser
+          filters:
+            tags:
+              only: /^[1-9]+.[0-9]+.[0-9]+.*/
```
doesn't it this way still only run for tags?

eunjae-lee replied:
https://github.com/algolia/algoliasearch-client-javascript/pull/1210/commits/5786f1afaf4f6ac63a04e1a6918c6aa883c3254f

aseure replied:
Yes my bad, I've left the filters we are using on all other API clients but in your case, you only need to build on `master`, since `shipjs trigger` already filters based on the commit message.

Perhaps the other filters are not needed anymore in that case (in all dependent jobs).

Haroenv commented on .circleci/config.yml:
```diff
@@ -119,27 +119,93 @@ jobs:
       - test_unit:
           version: '8'
+          name: 'test_unit_8'
+          filters:
```
I think most of these filters can also be removed

eunjae-lee replied:
https://github.com/algolia/algoliasearch-client-javascript/pull/1210/commits/3198d231ec0bb705f8d1d4f1b2df95aebf87d2e9

Haroenv reviewed (approved):
this looks good to me!

eunjae-lee commented:
@Haroenv Do you know what's going on here?
![image](https://user-images.githubusercontent.com/499898/94667112-a7870000-030e-11eb-8284-cb8c4e207ba0.png)

The job names seem to be messed up.

Haroenv commented:
the names of the tests have changed, so it's expected the "required" ones aren't ran. After we merge this PR we need to change which ones are required

eunjae-lee commented:
> the names of the tests have changed, so it's expected the "required" ones aren't ran. After we merge this PR we need to change which ones are required

Ah I didn't know the names have changed in this PR.
